xdsl/parser.py

```diff
@@ -1,1335 +1,1942 @@
+            return self.must_parse_function_type()
+
+    def must_parse_region_list(self) -> list[Region]:
+        """
+        Parses a sequence of regions for as long as there is a `{` in the input.
+        """
+        regions = []
+        while not self.tokenizer.is_eof() and self.tokenizer.starts_with("{"):
+            regions.append(self.must_parse_region())
+        return regions
+
+    # COMMON xDSL/MLIR code:
+    def must_parse_builtin_type_with_name(self, name: Span):
+        if name.text == "index":
+            return IndexType()
+        if (re_match := re.match(r"^[su]?i(\d+)$", name.text)) is not None:
+            signedness = {
+                "s": Signedness.SIGNED,
+                "u": Signedness.UNSIGNED,
+                "i": Signedness.SIGNLESS,
+            }
+            return IntegerType.from_width(int(re_match.group(1)),
+                                          signedness[name.text[0]])
+
+        if (re_match := re.match(r"^f(\d+)$", name.text)) is not None:
+            width = int(re_match.group(1))
+            type = {
+                16: Float16Type,
+                32: Float32Type,
+                64: Float64Type
+            }.get(width, None)
+            if type is None:
+                self.raise_error(
+                    "Unsupported floating point width: {}".format(width))
+            return type()
+
+        return self.must_parse_builtin_parametrized_type(name)
+
+    @abstractmethod
+    def must_parse_operation_details(
+        self,
+    ) -> tuple[list[Span], list[Span], dict[str, Attribute], list[Region],
+               FunctionType | None]:
+        """
+        Must return a tuple consisting of:
+            - a list of arguments to the operation
+            - a list of successor names
+            - the attributes attached to the OP
+            - the regions of the op
+            - An optional function type. If not supplied, must_parse_op_result_list must return a second value
+              containing the types of the returned SSAValues
+
+        Your implementation should make use of the following functions:
+            - must_parse_op_args_list
+            - must_parse_optional_attr_dict
+            - must_parse_
+        """
+        raise NotImplementedError()
 
-        # Register the SSA value names in the parser
-        for (idx, res) in enumerate(results):
-            if res[0] in self._ssaValues:
-                raise ParserError(start_pos,
-                                  f"SSA value {res[0]} is already defined")
-            self._ssaValues[res[0]] = op.results[idx]
-            if self.is_valid_name(res[0]):
-                self._ssaValues[res[0]].name = res[0]
+    @abstractmethod
+    def must_parse_op_args_list(self) -> list[Span]:
+        raise NotImplementedError()
 
-        return op
+    # HERE STARTS A SOMEWHAT CURSED COMPATIBILITY LAYER:
+    # since we don't want to rewrite all dialects currently, the new emulator needs to expose the same
```

Comment:
What emulator? Is this the connection between the old and the new parser?

Reply:
That is a typo, I meant parser :)